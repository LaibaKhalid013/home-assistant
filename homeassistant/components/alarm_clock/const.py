"""Constants for the Alarm Clock integration."""

import voluptuous as vol

from homeassistant.const import CONF_ICON, CONF_NAME
import homeassistant.helpers.config_validation as cv

DOMAIN = "alarm_clock"
ENTITY_ID_FORMAT = DOMAIN + ".{}"

ATTR_ALARM_TIME = "alarm_time"
ATTR_REPEAT_DAYS = "repeat_days"
ATTR_NEXT_ALARM = "next_alarm"

CONF_ALARM_TIME = "alarm_time"
CONF_REPEAT_DAYS = "repeat_days"

EVENT_ALARM_CLOCK_STARTED = "alarm_clock.started"
EVENT_ALARM_CLOCK_FINISHED = "alarm_clock.finished"
EVENT_ALARM_CLOCK_CANCELLED = "alarm_clock.cancelled"
EVENT_ALARM_CLOCK_CHANGED = "alarm_clock.changed"

STORAGE_KEY = DOMAIN
STORAGE_VERSION = 1

STORAGE_FIELDS = {
    vol.Required(CONF_NAME): cv.string,
    vol.Optional(CONF_ICON): cv.icon,
    vol.Required(CONF_ALARM_TIME): cv.time,
    vol.Required(CONF_REPEAT_DAYS): cv.weekdays,
}
