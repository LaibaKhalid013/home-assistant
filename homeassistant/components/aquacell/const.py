"""Constants for the Aquacell integration."""

DOMAIN = "aquacell"
DATA_AQUACELL = "DATA_AQUACELL"
