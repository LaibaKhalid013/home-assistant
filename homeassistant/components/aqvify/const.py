"""Constants for the Aqvify integration."""

import logging

DOMAIN = "aqvify"

CONF_DEVICE_KEY = "device_key"

LOGGER = logging.getLogger(__package__)
