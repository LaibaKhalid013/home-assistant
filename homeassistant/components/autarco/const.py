"""Constants for the Autarco integration."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Final

DOMAIN: Final = "autarco"
LOGGER = logging.getLogger(__package__)
SCAN_INTERVAL = timedelta(minutes=1)

CONF_PUBLIC_KEY: Final = "public_key"

SENSORS_SOLAR: Final = "solar"
