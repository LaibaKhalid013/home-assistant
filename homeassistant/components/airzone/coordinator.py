"""The Airzone integration."""

from __future__ import annotations

from asyncio import timeout
from datetime import timedelta
import logging
from typing import Any

from aioairzone.const import AZD_NEW_SYSTEMS, AZD_NEW_ZONES
from aioairzone.exceptions import AirzoneError
from aioairzone.localapi import AirzoneLocalApi

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import AIOAIRZONE_DEVICE_TIMEOUT_SEC, DOMAIN

SCAN_INTERVAL = timedelta(seconds=60)

_LOGGER = logging.getLogger(__name__)


class AirzoneUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
    """Class to manage fetching data from the Airzone device."""

    def __init__(self, hass: HomeAssistant, airzone: AirzoneLocalApi) -> None:
        """Initialize."""
        self.airzone = airzone

        super().__init__(
            hass,
            _LOGGER,
            name=DOMAIN,
            update_interval=SCAN_INTERVAL,
        )

    async def _async_check_new_devices(self, data: dict[str, Any]) -> None:
        """Reload entry if there are new devices."""
        if (
            len(data.get(AZD_NEW_SYSTEMS, [])) > 0
            or len(data.get(AZD_NEW_ZONES, [])) > 0
        ):
            assert self.config_entry
            self.hass.config_entries.async_schedule_reload(self.config_entry.entry_id)

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        async with timeout(AIOAIRZONE_DEVICE_TIMEOUT_SEC):
            try:
                await self.airzone.update()
            except AirzoneError as error:
                raise UpdateFailed(error) from error
            data = self.airzone.data()
        await self._async_check_new_devices(data)
        return data
