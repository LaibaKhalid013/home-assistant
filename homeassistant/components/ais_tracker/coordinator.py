"""AIS tracker coordinator."""

from __future__ import annotations

import asyncio
import logging
from socket import AF_INET, SO_REUSEPORT, SOCK_DGRAM, SOL_SOCKET, socket

from pyais.stream import SocketStream

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PORT
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
from homeassistant.util.json import json_loads_object

from .const import CONF_MMSIS, DOMAIN

LOGGER = logging.getLogger(__name__)

type AisTrackerConfigEntry = ConfigEntry[AisTrackerCoordinator]


class UDPReceiver(SocketStream):
    """Re-implementation of pyais.stream.UDPReceiver.

    can be removed, as soon as https://github.com/M0r13n/pyais/pull/148 is merged.
    """

    _fobj: socket

    def __init__(self, host: str, port: int) -> None:
        """Initialize the UDP receiver."""
        sock: socket = socket(AF_INET, SOCK_DGRAM)
        sock.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1)
        sock.bind((host, port))
        super().__init__(sock, preprocessor=None)

    def recv(self) -> bytes:
        """Receive data from socket."""
        return self._fobj.recvfrom(self.BUF_SIZE)[0]

    def close(self) -> None:
        """Close the UDP receiver."""
        self._fobj.close()


class AisTrackerCoordinator(
    DataUpdateCoordinator[dict[str, dict[str, float | int | str | None]]]
):
    """AIS tracker data update coordinator."""

    config_entry: AisTrackerConfigEntry

    def __init__(self, hass: HomeAssistant) -> None:
        """Initialize the AIS tracker coordinator."""
        super().__init__(
            hass,
            LOGGER,
            name=DOMAIN,
            update_interval=None,
        )
        self._receiver: UDPReceiver | None = None
        self._background_task: asyncio.Task | None = None
        self.data = {}

    async def async_setup(self) -> None:
        """Set up the AIS tracker coordinator."""

        self._receiver = UDPReceiver("", self.config_entry.data[CONF_PORT])

        async def async_ais_listerner():
            def ais_listerner():
                for msg_raw in self._receiver:
                    msg = json_loads_object(msg_raw.decode().to_json())
                    LOGGER.debug("received msg: %s", msg)
                    if (
                        msg.get("msg_type") in [1, 2, 3, 5]
                        and str(msg.get("mmsi")) in self.config_entry.data[CONF_MMSIS]
                    ):
                        self.hass.create_task(self.async_process_ais_message(msg))

            await self.hass.async_add_executor_job(ais_listerner)

        self._background_task = self.config_entry.async_create_background_task(
            hass=self.hass,
            target=async_ais_listerner(),
            name=f"{DOMAIN}_ais_listerner",
            eager_start=False,
        )

        self.config_entry.async_on_unload(self.async_teardown)

    async def async_teardown(self) -> None:
        """Tear down the AIS tracker coordinator."""
        if self._receiver is not None:
            self._receiver.close()
        if self._background_task is not None:
            self._background_task.cancel()

    async def async_process_ais_message(self, message: dict) -> None:
        """Process incoming ais message."""
        LOGGER.debug("process ais message: %s", message)
        if message.get("msg_type") in [1, 2, 3]:  # position reports
            data = {**self.data}
            data[str(message["mmsi"])] = message
            self.async_set_updated_data(data)

        elif message.get("msg_type") == 5:  # Static and voyage related data
            dreg = dr.async_get(self.hass)
            dreg.async_get_or_create(
                config_entry_id=self.config_entry.entry_id,
                identifiers={(DOMAIN, str(message["mmsi"]))},
                name=f"{message.get('shipname')} ({message.get('callsign')})",
                serial_number=str(message["mmsi"]),
                manufacturer="AIS tracker",
            )
