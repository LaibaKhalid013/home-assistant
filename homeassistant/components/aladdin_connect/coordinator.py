"""Define an object to coordinate fetching Aladdin Connect data."""

from datetime import timedelta
import logging

from genie_partner_sdk.client import AladdinConnectClient
from genie_partner_sdk.model import GarageDoor

from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


class AladdinConnectCoordinator(DataUpdateCoordinator[None]):
    """Aladdin Connect Data Update Coordinator."""

    doors: list[GarageDoor] | None = None

    def __init__(self, hass: HomeAssistant, acc: AladdinConnectClient) -> None:
        """Initialize."""
        super().__init__(
            hass,
            logger=_LOGGER,
            name=DOMAIN,
            update_interval=timedelta(seconds=15),
        )
        self.acc = acc

    async def async_setup(self) -> None:
        """Fetch initial data."""
        self.doors = await self.acc.get_doors()

    async def _async_update_data(self) -> None:
        """Fetch data from API endpoint."""
        assert self.doors is not None
        for door in self.doors:
            await self.acc.update_door(door.device_id, door.door_number)
