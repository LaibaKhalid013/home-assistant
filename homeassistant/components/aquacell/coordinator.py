"""Coordinator to update data from Aquacell API."""

import asyncio
from datetime import timedelta
import logging

from aioaquacell import (
    AquacellApi,
    AquacellApiException,
    AuthenticationFailed,
    Softener,
)

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryError
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import CONF_REFRESH_TOKEN, UPDATE_INTERVAL

_LOGGER = logging.getLogger(__name__)


class AquacellCoordinator(DataUpdateCoordinator[dict[str, Softener]]):
    """My aquacell coordinator."""

    config_entry: ConfigEntry

    def __init__(self, hass: HomeAssistant, aquacell_api: AquacellApi) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            _LOGGER,
            name="Aquacell Coordinator",
            update_interval=timedelta(seconds=UPDATE_INTERVAL),
        )

        self.refresh_token = self.config_entry.data[CONF_REFRESH_TOKEN]
        self.email = self.config_entry.data[CONF_EMAIL]
        self.password = self.config_entry.data[CONF_PASSWORD]
        self.aquacell_api = aquacell_api

    async def _async_update_data(self) -> dict[str, Softener]:
        """Fetch data from API endpoint.

        This is the place to pre-process the data to lookup tables
        so entities can quickly look up their data.
        """

        async with asyncio.timeout(10):
            _LOGGER.debug("Logging in using: %s", self.refresh_token)

            try:
                await self.aquacell_api.authenticate_refresh(self.refresh_token)
            except AuthenticationFailed as err:
                _LOGGER.debug(
                    "Authentication using refresh token failed due to: %s", err
                )
                _LOGGER.debug("Attempting to renew refresh token")
                refresh_token = await self.aquacell_api.authenticate(
                    self.email, self.password
                )
                self.refresh_token = refresh_token
                self._update_config_entry_refresh_token()

            _LOGGER.debug("Logged in, new token: %s", self.aquacell_api.id_token)
        try:
            softeners = await self.aquacell_api.get_all_softeners()
        except AuthenticationFailed as err:
            raise ConfigEntryError from err
        except AquacellApiException as err:
            raise UpdateFailed(f"Error communicating with API: {err}") from err

        return {softener.dsn: softener for softener in softeners}

    def _update_config_entry_refresh_token(self) -> None:
        """Update or delete the refresh_token in the Config Entry."""
        data = {
            **self.config_entry.data,
            CONF_REFRESH_TOKEN: self.refresh_token,
        }

        self.hass.config_entries.async_update_entry(self.config_entry, data=data)
