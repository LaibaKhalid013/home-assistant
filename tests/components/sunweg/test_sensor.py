"""Tests for the Sun WEG sensor."""

from freezegun.api import FrozenDateTimeFactory
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.sunweg.const import DOMAIN
from homeassistant.components.sunweg.coordinator import SCAN_INTERVAL
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component

from .common import SUNWEG_MOCK_ENTRY

from tests.common import async_fire_time_changed


async def test_sensor_total(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    freezer: FrozenDateTimeFactory,
    api_fixture,
    plant_fixture_alternative,
) -> None:
    """Test sensor type total."""
    mock_entry = SUNWEG_MOCK_ENTRY
    mock_entry.add_to_hass(hass)

    assert await async_setup_component(hass, DOMAIN, mock_entry.data)
    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.plant_123_total_money_lifetime")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.plant_123_total_energy_today")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.plant_123_total_output_power")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.plant_123_total_lifetime_energy_output")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.plant_123_total_last_update")
    assert state.state == snapshot
    assert state.attributes == snapshot

    api_fixture["plant"].return_value = plant_fixture_alternative

    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.plant_123_total_last_update")
    assert state.state == snapshot
    assert state.attributes == snapshot


async def test_sensor_inverter(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    freezer: FrozenDateTimeFactory,
    api_fixture,
    inverter_fixture_invalid,
    plant_fixture,
) -> None:
    """Test sensor type inverter."""
    mock_entry = SUNWEG_MOCK_ENTRY
    mock_entry.add_to_hass(hass)

    assert await async_setup_component(hass, DOMAIN, mock_entry.data)
    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.inversor01_energy_today")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_lifetime_energy_output")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_ac_frequency")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_output_power")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_temperature")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_power_factor")
    assert state.state == snapshot
    assert state.attributes == snapshot

    plant = plant_fixture
    plant.inverters.clear()
    plant.inverters.append(inverter_fixture_invalid)

    api_fixture["plant"].return_value = plant

    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.inversor01_energy_today")
    assert state.state == snapshot
    assert state.attributes == snapshot


async def test_sensor_phase(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    freezer: FrozenDateTimeFactory,
    api_fixture,
    plant_fixture_alternative,
) -> None:
    """Test sensor type phase."""
    mock_entry = SUNWEG_MOCK_ENTRY
    mock_entry.add_to_hass(hass)

    assert await async_setup_component(hass, DOMAIN, mock_entry.data)
    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.inversor01_phasea_voltage")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_phasea_amperage")
    assert state.state == snapshot
    assert state.attributes == snapshot

    api_fixture["plant"].return_value = plant_fixture_alternative

    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()

    state = hass.states.get("sensor.inversor01_phasea_voltage")
    assert state.state == snapshot
    assert state.attributes == snapshot


async def test_sensor_string(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    freezer: FrozenDateTimeFactory,
    api_fixture,
    plant_fixture_alternative,
) -> None:
    """Test sensor type string."""
    mock_entry = SUNWEG_MOCK_ENTRY
    mock_entry.add_to_hass(hass)

    assert await async_setup_component(hass, DOMAIN, mock_entry.data)
    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done(wait_background_tasks=True)

    state = hass.states.get("sensor.inversor01_str1_voltage")
    assert state.state == snapshot
    assert state.attributes == snapshot

    state = hass.states.get("sensor.inversor01_str1_amperage")
    assert state.state == snapshot
    assert state.attributes == snapshot

    api_fixture["plant"].return_value = plant_fixture_alternative

    freezer.tick(SCAN_INTERVAL)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    state = hass.states.get("sensor.inversor01_str1_voltage")
    assert state.state == snapshot
    assert state.attributes == snapshot
