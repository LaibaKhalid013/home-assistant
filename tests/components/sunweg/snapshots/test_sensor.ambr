# serializer version: 1
# name: test_sensor_inverter
  '0.0'
# ---
# name: test_sensor_inverter.1
  ReadOnlyDict({
    'device_class': 'energy',
    'friendly_name': 'INVERSOR01 Energy today',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    'unit_of_measurement': <UnitOfEnergy.KILO_WATT_HOUR: 'kWh'>,
  })
# ---
# name: test_sensor_inverter.10
  '0'
# ---
# name: test_sensor_inverter.11
  ReadOnlyDict({
    'friendly_name': 'INVERSOR01 Power Factor',
    'icon': 'mdi:solar-power',
  })
# ---
# name: test_sensor_inverter.12
  'unknown'
# ---
# name: test_sensor_inverter.13
  ReadOnlyDict({
    'device_class': 'energy',
    'friendly_name': 'INVERSOR01 Energy today',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    'unit_of_measurement': 'None',
  })
# ---
# name: test_sensor_inverter.2
  '0.0'
# ---
# name: test_sensor_inverter.3
  ReadOnlyDict({
    'device_class': 'energy',
    'friendly_name': 'INVERSOR01 Lifetime energy output',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.TOTAL: 'total'>,
    'unit_of_measurement': 'MWh',
  })
# ---
# name: test_sensor_inverter.4
  '0.0'
# ---
# name: test_sensor_inverter.5
  ReadOnlyDict({
    'device_class': 'frequency',
    'friendly_name': 'INVERSOR01 AC frequency',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfFrequency.HERTZ: 'Hz'>,
  })
# ---
# name: test_sensor_inverter.6
  '1'
# ---
# name: test_sensor_inverter.7
  ReadOnlyDict({
    'device_class': 'power',
    'friendly_name': 'INVERSOR01 Output power',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
    'unit_of_measurement': 'kW',
  })
# ---
# name: test_sensor_inverter.8
  '23.2'
# ---
# name: test_sensor_inverter.9
  ReadOnlyDict({
    'device_class': 'temperature',
    'friendly_name': 'INVERSOR01 Temperature',
    'icon': 'mdi:temperature-celsius',
    'unit_of_measurement': <UnitOfTemperature.CELSIUS: 'Â°C'>,
  })
# ---
# name: test_sensor_phase
  '120.0'
# ---
# name: test_sensor_phase.1
  ReadOnlyDict({
    'device_class': 'voltage',
    'friendly_name': 'INVERSOR01 PhaseA Voltage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricPotential.VOLT: 'V'>,
  })
# ---
# name: test_sensor_phase.2
  '3.2'
# ---
# name: test_sensor_phase.3
  ReadOnlyDict({
    'device_class': 'current',
    'friendly_name': 'INVERSOR01 PhaseA Amperage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricCurrent.AMPERE: 'A'>,
  })
# ---
# name: test_sensor_phase.4
  'unknown'
# ---
# name: test_sensor_phase.5
  ReadOnlyDict({
    'device_class': 'voltage',
    'friendly_name': 'INVERSOR01 PhaseA Voltage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricPotential.VOLT: 'V'>,
  })
# ---
# name: test_sensor_string
  '450.3'
# ---
# name: test_sensor_string.1
  ReadOnlyDict({
    'device_class': 'voltage',
    'friendly_name': 'INVERSOR01 STR1 Voltage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricPotential.VOLT: 'V'>,
  })
# ---
# name: test_sensor_string.2
  '23.4'
# ---
# name: test_sensor_string.3
  ReadOnlyDict({
    'device_class': 'current',
    'friendly_name': 'INVERSOR01 STR1 Amperage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricCurrent.AMPERE: 'A'>,
  })
# ---
# name: test_sensor_string.4
  'unknown'
# ---
# name: test_sensor_string.5
  ReadOnlyDict({
    'device_class': 'voltage',
    'friendly_name': 'INVERSOR01 STR1 Voltage',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfElectricPotential.VOLT: 'V'>,
  })
# ---
# name: test_sensor_total
  '12.786912'
# ---
# name: test_sensor_total.1
  ReadOnlyDict({
    'friendly_name': 'Plant #123 Total Money lifetime',
    'icon': 'mdi:cash',
    'unit_of_measurement': 'R$',
  })
# ---
# name: test_sensor_total.10
  'unknown'
# ---
# name: test_sensor_total.11
  ReadOnlyDict({
    'device_class': 'date',
    'friendly_name': 'Plant #123 Total Last Update',
    'icon': 'mdi:solar-power',
  })
# ---
# name: test_sensor_total.12
  'unknown'
# ---
# name: test_sensor_total.13
  ReadOnlyDict({
    'device_class': 'date',
    'friendly_name': 'Plant #123 Total Last Update',
    'icon': 'mdi:solar-power',
  })
# ---
# name: test_sensor_total.2
  '24.0'
# ---
# name: test_sensor_total.3
  ReadOnlyDict({
    'device_class': 'energy',
    'friendly_name': 'Plant #123 Total Energy Today',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
    'unit_of_measurement': <UnitOfEnergy.KILO_WATT_HOUR: 'kWh'>,
  })
# ---
# name: test_sensor_total.4
  '29.5'
# ---
# name: test_sensor_total.5
  ReadOnlyDict({
    'device_class': 'power',
    'friendly_name': 'Plant #123 Total Output Power',
    'icon': 'mdi:solar-power',
    'unit_of_measurement': <UnitOfPower.KILO_WATT: 'kW'>,
  })
# ---
# name: test_sensor_total.6
  '332.2'
# ---
# name: test_sensor_total.7
  ReadOnlyDict({
    'device_class': 'energy',
    'friendly_name': 'Plant #123 Total Lifetime energy output',
    'icon': 'mdi:solar-power',
    'state_class': <SensorStateClass.TOTAL: 'total'>,
    'unit_of_measurement': <UnitOfEnergy.KILO_WATT_HOUR: 'kWh'>,
  })
# ---
# name: test_sensor_total.8
  '2023-02-16T14:22:37'
# ---
# name: test_sensor_total.9
  ReadOnlyDict({
    'device_class': 'date',
    'friendly_name': 'Plant #123 Total Last Update',
    'icon': 'mdi:solar-power',
  })
# ---
